server/src/main/java/io/crate/expression/scalar/HasDatabasePrivilegeFunction.java

```diff
@@ -0,0 +1,190 @@
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.function.BiFunction;
+
+import javax.annotation.Nullable;
+
+import org.elasticsearch.common.TriFunction;
+
+import io.crate.Constants;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+import io.crate.user.Privilege;
+import io.crate.user.User;
+import io.crate.user.UserLookup;
+
+public class HasDatabasePrivilegeFunction extends HasPrivilegeFunction {
+
+    public static final String NAME = "has_database_privilege";
+
+    private static final TriFunction<User, Object, Collection<Privilege.Type>, Boolean> CHECK_BY_DB_NAME =
+        (user, db, privileges) -> {
+            if (Constants.DB_NAME.equals(db) == false) {
+                throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                                                 "database \"%s\" does not exist",
+                                                                 db));
+            }
+            return checkPrivileges(user, privileges);
+        };
+
+    private static final TriFunction<User, Object, Collection<Privilege.Type>, Boolean> CHECK_BY_DB_OID =
+        (user, db, privileges) -> {
+            if (Constants.DB_OID != (Integer) db) {
+                throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                                                 "database with OID \"%d\" does not exist",
+                                                                 db));
+            }
+            return checkPrivileges(user, privileges);
+        };
+
+    private static boolean checkPrivileges(User user, Collection<Privilege.Type> privileges) {
+        if (privileges.contains(Privilege.Type.DQL)) { // CONNECT
+            return true;
+        }
+
+        boolean result = true;
+        if (privileges.contains(Privilege.Type.DML)) { // TEMP privilege
+            result = false;
+        }
+        if (privileges.contains(Privilege.Type.DDL)) { // CREATE privilege
+            result = hasCreatePrivilege(user);
+        }
+        return result;
+    }
+
+    private static boolean hasCreatePrivilege(User user) {
+        for (Privilege p : user.privileges()) {
+            if (p.ident().type() == Privilege.Type.DDL &&
+                (p.ident().clazz() == Privilege.Clazz.SCHEMA || p.ident().clazz() == Privilege.Clazz.CLUSTER)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Nullable
+    protected Collection<Privilege.Type> parsePrivileges(String privilege) {
+        Collection<Privilege.Type> toCheck = new HashSet<>();
+        String[] privileges = privilege.toLowerCase(Locale.ENGLISH).split(",");
+        for (String p : privileges) {
+            p = p.trim();
+            switch (p) {
+                case "connect" -> toCheck.add(Privilege.Type.DQL);
```

Comment:
Don't know how this is handled internally, but in theory a user that has no explicit privileges granted, still has access to some `information_schema` and `pg_catalog` tables and therefore could connect (and query) the database

Reply:
This is a mapping we do ourselves, there is no connect privilege for CrateDB, so mapping it to DQL, Also mapping the rest `TEMP`, and `CREATE` to the available `DML` and `DDL` enum values we have.
When we validate here: https://github.com/crate/crate/pull/13377/files#diff-2328b4da2125364630421b4cb3da5089ba02a5638fd0e83b1520740b07786099R65 we check for DQL => connect, if the user exists it's true, for `DML` => temp always false, we don't have temp tables in CrateDB, for `DDL`=> create, we check AL on cluster of DDL on any schema.